### PR TITLE
Added ability for vSphere to reconnect on secret update

### LINKED ISSUE #90844: PVCs can still be provisioned after the password has been changed vSphere
**What happened**:

When the password is changed in the vSphere secret for the user that is configured for the cluster, storage can still be provisioned despite having a password that is no longer valid. 

**What you expected to happen**:

 PVC provisioning should fail with "Cannot complete login due to an incorrect user name or password."

**How to reproduce it (as minimally and precisely as possible)**:

1. Deploy a cluster with vSphere integration. Check that dynamically provisioning storage works.
2. Change the user's pasword in referenced vSphere secret
3. Provision a new PVC, it will be created.

**Anything else we need to know?**:

 Even 24 hours after the password change in vSphere, OCP can still provision storage.
 Changing the password in the referenced secret `vsphere-creds` does not make any change. You can change it to another password that is not correct, and storage will still be provisioned.

 Manually deleting the kube-apiserver, kube-controller-manager pods does not make a difference. 

**Environment**:
- Kubernetes version (use `kubectl version`): v1.16.2
- Cloud provider or hardware configuration: vSphere
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

While vSphere provider allows to specify secret with credentials, used to access vCenter resources, changing credentials in that secret would not refresh connection. This PR intends to fix that.

**Which issue(s) this PR fixes**:

Fixes #90844 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Provider-specific Notes: vsphere: vsphere.conf - new option to disable credentials secret management for performance concerns
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by enxebre** [comment]
This seems reasonable to me. Friendly pinging some maintainers @abrarshivani @frapposelli @akutz

**Comment by Danil-Grigorev** [comment]
@frapposelli I've updated PR as it was failing tests for some reason. The CI looks good now, may I get another approval? Thanks

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1484]: What's the reason for refreshing on delete?

**Review by liggitt — COMMENTED** [review]
Iterating over all nodes to propagate a credential change (especially without retry) seems expensive and fragile.

It looks like SecretCredentialManager already has some refresh logic in updateCredentialsMap which is called from GetCredential().

The individual node connections appear to update their credentials (`vsphereInstance.conn.UpdateCredentials`) if needed whenever vcConnect is called. Is that not working properly?

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1494]: This check and the one above should ideally happen in the event handler before calling refreshNodesForSecret.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (Danil-Grigorev) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1484]: The nature of the issue basically describes it. If the secret is removed, the results should be equal to situation when a `vsphere_password` env variable was removed from the config - refresh all the nodes state, and trigger events related to a misconfiguration

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (Danil-Grigorev) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1499]: Yeah, agree, will give it a dedicated method.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (andrewsykim) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1484]: Having a hard time seeing how the behavior of `GetNodeDetails` changes when this secret is deleted.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (Danil-Grigorev) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1470]: That's the issue faced by QE at some point. If some of the configuration is dedicated to the secret content, and there is another manager resource updating cluster configuration on config's content change, this part was left over, as nothing would recognize the secret being `added/updated/deleted`

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (Danil-Grigorev) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1494]: Agree, just happened to see 3 occurrences, so merged it into single location.

**Comment by Danil-Grigorev** [comment]
> Iterating over all nodes to propagate a credential change (especially without retry) seems expensive and fragile.

It is actually simpler that that (with the latest change) as it is using cached node objects for that reason.

> It looks like SecretCredentialManager already has some refresh logic in updateCredentialsMap which is called from GetCredential().
> The individual node connections appear to update their credentials 
(`vsphereInstance.conn.UpdateCredentials`) if needed whenever vcConnect is called. Is that not working properly?

Yes, but assuming the case that this is a kubernetes object, this secret it's not serving a purpose of matching a the state of the world. Whenever you restart daemons manually, it goes through the full process again, but this change is manual. The intention for this PR is to delegate the part about renewing connection to VMs (and potentially finishing a failed installation procedure) same way, as it done for Node object changes now.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (Danil-Grigorev) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1531]: Having and error here means that it failed during couple of retries while trying to (re)discover node, and update VM state, so this logic is already handled.

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (Danil-Grigorev) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go]: It should fail at this point with the following error - https://github.com/Danil-Grigorev/kubernetes/blob/8a31a8921350776f0077ea17da1dbbaccd42ec35/staging/src/k8s.io/legacy-cloud-providers/vsphere/credentialmanager.go#L95

Comment (Danil-Grigorev) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go]: Aaactually I think you are right. Having an additional event listener just to throw a log error message does not worth it :)

**Comment by Danil-Grigorev** [comment]
@andrewsykim  @liggitt Thanks for review, may I get another one? I've removed unnecessary dependency on node lister, and restricted reconciles only on secret credentials change, or secret being added. There is a little code restructuring in there to remuve duplicated parts + now it won't appear that an undesirable side-effect being used to update VM states for cached nodes. Deleting secret will be ignored.

@frapposelli May I get another lgtm in case the code is ok?

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (liggitt) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1472]: don't log content here, can log the type instead with:

```go
utilruntime.HandleError(fmt.Errorf("expected v1.Secret, got %T", obj))
```

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go
Comment (liggitt) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:1489]: don't log content here, can log the type instead with:

```go
utilruntime.HandleError(fmt.Errorf("expected v1.Secret, got %T", obj))
```

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/nodemanager.go
Comment (liggitt) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/nodemanager.go:313]: short-circuiting here means if there are 10 nodes, and the first one errors, the remaining 9 are left with their existing connections, right? doesn't that mean that https://github.com/kubernetes/kubernetes/issues/90844 is still not fixed?

**Comment by liggitt** [comment]
I did another pass on the PR and noted some issues with logging secret content, and what appears to me to still be an issue with not handling errors properly.

I'm not very familiar with the in-tree vsphere cloud provider, and find the way it manages nodes pretty difficult to follow, so I'll defer to the vsphere maintainers on whether this is a good approach, but I question whether hooking up a secret watcher directly to something that re-establishes connections to all nodes in the cluster will scale well on a multi-thousand-node cluster.

**Comment by dvonthenen** [comment]
> I did another pass on the PR and noted some issues with logging secret content, and what appears to me to still be an issue with not handling errors properly.
> 
> I'm not very familiar with the in-tree vsphere cloud provider, and find the way it manages nodes pretty difficult to follow, so I'll defer to the vsphere maintainers on whether this is a good approach, but I question whether hooking up a secret watcher directly to something that re-establishes connections to all nodes in the cluster will scale well on a multi-thousand-node cluster.

The secret watcher is required on each node to support certain functionality such as zones and etc. There is active code running (obviously on the master nodes for controller portion), but also active code running on all worker nodes which requires vCenter creds and therefore you have a watcher on every node in the cluster.

With that many watchers on every node in the cluster you are bound to have performance issues (like on etcd) on a multi-thousand node cluster. There should definitely be some ability to disable this watcher functionality for situations where workers nodes in the cluster don't need credentials because the provider isn't using features that require it.

**Comment by Danil-Grigorev** [comment]
@dvonthenen Please review, I added an option in the config file to disable secret watcher

**Comment by dvonthenen** [comment]
the new option seems ok with me!

**Comment by Danil-Grigorev** [comment]
Folks, may I get an approval to close the issue? @abrarshivani @frapposelli @divyenpatel @dougm

**Comment by dvonthenen** [comment]
/lgtm

but then again I only work on the external cloud provider code and I don't have approval access

**Review by divyenpatel — APPROVED** [review]

**Comment by divyenpatel** [comment]
@Danil-Grigorev Can you fix bazel-build error.

> ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/staging/src/k8s.io/legacy-cloud-providers/vsphere/BUILD:51:1: no such package 'vendor/k8s.io/klog': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.

**Comment by enxebre** [comment]
@Danil-Grigorev path sould be -> `"//vendor/k8s.io/klog/v2:go_default_library",`

**Comment by Danil-Grigorev** [comment]
> @Danil-Grigorev Can you fix bazel-build error.
> 
> > ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/staging/src/k8s.io/legacy-cloud-providers/vsphere/BUILD:51:1: no such package 'vendor/k8s.io/klog': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.

Fixed now.

**Comment by Danil-Grigorev** [comment]
@divyenpatel CI is passing, ready to merge?

**Pull request merged**
